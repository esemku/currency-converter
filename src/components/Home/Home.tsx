import { AmountInput, Button, SelectCurrency } from 'components/common';
import { IconLeftRightArrows, IconUpDownArrows } from 'components/common/icons';
import React, { useEffect, useState } from 'react';
import { shallowEqual, useDispatch, useSelector } from 'react-redux';
import {
  convertFromCurrencyRequest,
  getCurrenciesListRequest,
} from 'redux/actions/currencyConverterActions';
import {
  getConvertedAmount,
  getCurrenciesList,
} from 'redux/selectors/currenciesSelectors';
import { ICurrencyListItem } from 'types/currencies';
import useStyles from './styles';

const Home: React.FC = () => {
  const styles = useStyles();
  const [amountInput, setAmountInput] = useState(null);
  const [currency, setCurrency] = useState({
    from: 'PLN',
    to: 'GBP',
  });
  const dispatch = useDispatch();
  const { currenciesList } = useSelector(getCurrenciesList, shallowEqual);
  const { convertedAmount } = useSelector(getConvertedAmount, shallowEqual);
  const { convertedFrom } = useSelector(getConvertedAmount, shallowEqual);
  const { convertedTo } = useSelector(getConvertedAmount, shallowEqual);

  useEffect(() => {
    dispatch(getCurrenciesListRequest());
  }, []);

  const convertCurrencies = () => {
    dispatch(
      convertFromCurrencyRequest({
        fromCurrency: currency.from,
        toCurrency: currency.to,
        amount: Number(amountInput),
      }),
    );
  };

  const handleCurrencyFrom = (e: React.ChangeEvent<HTMLSelectElement>) => {
    setCurrency({
      ...currency,
      from: e.target.value,
    });
  };

  const handleCurrencyTo = (e: React.ChangeEvent<HTMLSelectElement>) => {
    setCurrency({
      ...currency,
      to: e.target.value,
    });
  };

  const exchangeCurrencySelectValues = () => {
    setCurrency({
      from: currency.to,
      to: currency.from,
    });
  };

  const getCurrencyName = (currencyCode: string) => {
    return currenciesList.filter(
      (currencyObject: ICurrencyListItem) =>
        currencyObject.iso === currencyCode,
    );
  };

  const formatConvertedAmount = (amount: number) => {
    const arr = String(amount).split('.');

    return (
      <h2 className={styles.convertedAmount}>
        {`${arr[0]}.${arr[1].substring(0, 2)}`}
        <span className={styles.convertedAmountGloom}>
          {arr[1].substring(2, 4)}&nbsp;
        </span>
        {getCurrencyName(convertedTo)[0].currency_name}
      </h2>
    );
  };

  return (
    <div className={styles.root}>
      <div className={styles.container}>
        <h3 className={styles.contentHeader}>Currency Converter</h3>
        <div className={styles.dashedLine} />
        <div className={styles.inputsWrapper}>
          <AmountInput
            inputName="amount"
            label="Amount"
            onChange={setAmountInput}
          />
          <SelectCurrency
            id="currencyFrom"
            label="From"
            currenciesList={currenciesList}
            value={currency.from}
            onChange={handleCurrencyFrom}
          />
          <IconUpDownArrows
            onClick={exchangeCurrencySelectValues}
            className={styles.iconUpDownArrows}
          />
          <IconLeftRightArrows
            onClick={exchangeCurrencySelectValues}
            className={styles.iconLeftRightArrows}
          />
          <SelectCurrency
            id="currencyTo"
            label="To"
            currenciesList={currenciesList}
            value={currency.to}
            onChange={handleCurrencyTo}
          />
        </div>
        <div className={styles.converterResultWrapper}>
          {Array.isArray(currenciesList) &&
            currenciesList.length > 0 &&
            convertedAmount && (
              <div data-testid="convertedAmountWrapper">
                <span
                  className={styles.resultTextCurrencyToConvert}
                >{`${amountInput} ${
                  getCurrencyName(convertedFrom)[0].currency_name
                } =`}</span>
                {formatConvertedAmount(convertedAmount)}
              </div>
            )}
          <Button
            name="Convert"
            onClick={convertCurrencies}
            disabled={!amountInput}
            className={styles.convertButton}
          />
        </div>
      </div>
    </div>
  );
};

export default Home;
