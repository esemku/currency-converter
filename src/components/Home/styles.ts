import { makeStyles } from '@material-ui/core';
import { colors, fonts } from 'theme';

export default makeStyles((theme: any) => ({
  root: {
    padding: 16,
    [theme.breakpoints.up('sm')]: {
      padding: 50,
    },
  },
  inputsWrapper: {
    [theme.breakpoints.up('md')]: {
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'space-between',
    },
  },
  container: {
    width: '100%',
    backgroundColor: colors.grayQuinary,
    borderRadius: theme.shape.borderRadius,
    padding: '8px 16px 16px',
    [theme.breakpoints.up('sm')]: {
      padding: '16px 40px',
    },
  },
  contentHeader: {
    margin: '0px 0px 10px',
  },
  dashedLine: {
    width: '100%',
    borderBottomStyle: 'dashed',
    borderBottom: `1px solid ${colors.graySenary}`,
    marginBottom: 16,
  },
  converterWrapper: {
    backgroundColor: colors.graySecondary,
    borderRadius: theme.shape.borderRadius,
    padding: '26px 30px',
  },
  resultTextCurrencyToConvert: {
    fontSize: fonts.size.mobile.span,
    fontWeight: fonts.weight.mobile.span,
    lineHeight: fonts.lineHeight.mobile.span,
    [theme.breakpoints.up('md')]: {
      fontSize: fonts.size.desktop.span,
      fontWeight: fonts.weight.desktop.span,
      lineHeight: fonts.lineHeight.desktop.span,
      whiteSpace: 'nowrap',
    },
    color: colors.grayQuaternary,
  },
  convertedAmount: {
    fontSize: fonts.size.mobile.h2,
    fontWeight: fonts.weight.mobile.h2,
    lineHeight: fonts.lineHeight.mobile.h2,
    [theme.breakpoints.up('md')]: {
      fontSize: fonts.size.desktop.h2,
      fontWeight: fonts.weight.desktop.h2,
      lineHeight: fonts.lineHeight.desktop.h2,
      whiteSpace: 'nowrap',
      margin: 0,
    },
    color: colors.whitePrimary,
    margin: '8px 0',
  },
  convertedAmountGloom: {
    opacity: 0.5,
  },
  convertButton: {
    marginTop: 16,
    [theme.breakpoints.up('md')]: {
      marginTop: 0,
    },
  },
  iconUpDownArrows: {
    [theme.breakpoints.up('md')]: {
      display: 'none',
    },
  },
  iconLeftRightArrows: {
    display: 'none',
    marginTop: 4,
    [theme.breakpoints.up('md')]: {
      display: 'block',
    },
  },
  converterResultWrapper: {
    [theme.breakpoints.up('md')]: {
      display: 'flex',
      flexFlow: 'row-reverse',
      alignItems: 'center',
      justifyContent: 'flex-end',
      minHeight: 51,
    },
  },
}));
