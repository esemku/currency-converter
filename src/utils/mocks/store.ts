export const store = {
  currencies: {
    currenciesList: [
      { iso: 'PLN', currency_name: 'Pol', is_obsolete: false },
      { iso: 'GBP', currency_name: 'Eng', is_obsolete: false },
      { iso: 'NOK', currency_name: 'Nor', is_obsolete: false },
      { iso: 'CAD', currency_name: 'Can', is_obsolete: false },
    ],
    convertedAmount: {
      amount: 5000,
      confertedFrom: 'PLN',
      convertedTo: 'GBP',
    },
  },
};
