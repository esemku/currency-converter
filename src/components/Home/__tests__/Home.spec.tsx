import { fireEvent, render, screen, within } from '@testing-library/react';
import React from 'react';
import * as ReactRedux from 'react-redux';
import configureStore from 'redux-mock-store';
import * as MockStore from 'utils/mocks/store';
import Home from '../Home';

describe('<Home />', () => {
  const mockStore = configureStore();
  const store = mockStore(MockStore.store);

  it('renders correctly', () => {
    const { container } = render(
      <ReactRedux.Provider store={store}>
        <Home />
      </ReactRedux.Provider>,
    );

    expect(container).toMatchSnapshot();
  });

  it('unlocking the convert button after entering amount', () => {
    const { getByTestId, queryByTestId } = render(
      <ReactRedux.Provider store={store}>
        <Home />
      </ReactRedux.Provider>,
    );

    expect(getByTestId('convertButton')).toHaveAttribute('disabled');

    fireEvent.change(screen.getByTestId('amountInput'), {
      target: { value: 5000 },
    });
    expect(getByTestId('convertButton')).not.toHaveAttribute('disabled');
    expect(queryByTestId('convertedAmountWrapper')).not.toBeInTheDocument();
  });

  it('changes value of currencyFrom select properly', () => {
    const { getByTestId, getByRole, queryByText } = render(
      <ReactRedux.Provider store={store}>
        <Home />
      </ReactRedux.Provider>,
    );

    const fromCurrencySelect = within(
      getByTestId('selectCurrencycurrencyFrom'),
    ).getByRole('button');

    fireEvent.mouseDown(fromCurrencySelect);
    const listbox = within(getByRole('listbox'));
    fireEvent.mouseDown(listbox.getByText('NOK'));

    expect(queryByText('NOK')).toBeInTheDocument();
  });

  it('changes value of currencyTo select properly', () => {
    const { getByTestId, getByRole, queryByText } = render(
      <ReactRedux.Provider store={store}>
        <Home />
      </ReactRedux.Provider>,
    );

    const toCurrencySelect = within(
      getByTestId('selectCurrencycurrencyTo'),
    ).getByRole('button');

    fireEvent.mouseDown(toCurrencySelect);
    const listbox = within(getByRole('listbox'));
    fireEvent.mouseDown(listbox.getByText('CAD'));

    expect(queryByText('CAD')).toBeInTheDocument();
  });
});
