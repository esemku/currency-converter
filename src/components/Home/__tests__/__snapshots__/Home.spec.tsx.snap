// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> renders correctly 1`] = `
<div>
  <div
    class="makeStyles-root-1"
  >
    <div
      class="makeStyles-container-3"
    >
      <h3
        class="makeStyles-contentHeader-4"
      >
        Currency Converter
      </h3>
      <div
        class="makeStyles-dashedLine-5"
      />
      <div
        class="makeStyles-inputsWrapper-2"
      >
        <div
          class="makeStyles-root-14"
        >
          <label
            class="makeStyles-label-16"
            for="amount"
          >
            Amount
          </label>
          <input
            class="makeStyles-input-15"
            data-testid="amountInput"
            id="amount"
            inputmode="decimal"
            name="amount"
            placeholder="e.g.  1,000.00"
            type="text"
            value=""
          />
        </div>
        <div
          class="makeStyles-root-17"
        >
          <label
            class="makeStyles-label-21"
            for="currencyFrom"
          >
            From
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline makeStyles-select-18"
            data-testid="selectCurrencycurrencyFrom"
          >
            <div
              aria-haspopup="listbox"
              aria-labelledby="currencyFrom"
              class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
              id="currencyFrom"
              role="button"
              tabindex="0"
            >
              <span
                class="sc-bdfBQB sc-gsTEea jczEZi fzESjn makeStyles-currencyFlag-22"
              />
              PLN
              <span
                class="makeStyles-currencyName-23"
              >
                 - 
                Pol
              </span>
            </div>
            <input
              aria-hidden="true"
              class="MuiSelect-nativeInput"
              tabindex="-1"
              value="PLN"
            />
            <svg
              class="MuiSelect-icon"
              fill="none"
              height="20"
              style="margin-top: 1px;"
              viewBox="0 0 24 24"
              width="40"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M11.434 12.734L7.25 8.55a1.06 1.06 0 00-1.5 1.5l5.543 5.543a1 1 0 001.414 0l5.543-5.543a1.06 1.06 0 10-1.5-1.5l-4.184 4.184a.8.8 0 01-1.132 0z"
                fill="#565674"
              />
            </svg>
          </div>
        </div>
        <svg
          class="makeStyles-root-24 makeStyles-iconUpDownArrows-11"
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M7 6.6V20c0 .6.4 1 1 1s1-.4 1-1V6.6H7zm8 10.8V4c0-.6.4-1 1-1s1 .4 1 1v13.4h-2z"
            fill="#000"
          />
          <path
            d="M3 6.6h10L8.7 2.3c-.4-.4-1-.4-1.4 0L3 6.6zm8 10.8h10l-4.3 4.3c-.4.4-1 .4-1.4 0L11 17.4z"
            fill="#000"
            opacity="0.8"
          />
        </svg>
        <svg
          class="makeStyles-root-25 makeStyles-iconLeftRightArrows-12"
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M17.4 7H4c-.6 0-1 .4-1 1s.4 1 1 1h13.4V7zM6.6 15H20c.6 0 1 .4 1 1s-.4 1-1 1H6.6v-2z"
            fill="#000"
          />
          <path
            d="M17.4 3v10l4.3-4.3c.4-.4.4-1 0-1.4L17.4 3zM6.6 11v10l-4.3-4.3c-.4-.4-.4-1 0-1.4L6.6 11z"
            fill="#000"
            opacity="0.3"
          />
        </svg>
        <div
          class="makeStyles-root-17"
        >
          <label
            class="makeStyles-label-21"
            for="currencyTo"
          >
            To
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline makeStyles-select-18"
            data-testid="selectCurrencycurrencyTo"
          >
            <div
              aria-haspopup="listbox"
              aria-labelledby="currencyTo"
              class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
              id="currencyTo"
              role="button"
              tabindex="0"
            >
              <span
                class="sc-bdfBQB sc-gsTEea jczEZi eqTMSR makeStyles-currencyFlag-22"
              />
              GBP
              <span
                class="makeStyles-currencyName-23"
              >
                 - 
                Eng
              </span>
            </div>
            <input
              aria-hidden="true"
              class="MuiSelect-nativeInput"
              tabindex="-1"
              value="GBP"
            />
            <svg
              class="MuiSelect-icon"
              fill="none"
              height="20"
              style="margin-top: 1px;"
              viewBox="0 0 24 24"
              width="40"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M11.434 12.734L7.25 8.55a1.06 1.06 0 00-1.5 1.5l5.543 5.543a1 1 0 001.414 0l5.543-5.543a1.06 1.06 0 10-1.5-1.5l-4.184 4.184a.8.8 0 01-1.132 0z"
                fill="#565674"
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        class="makeStyles-converterResultWrapper-13"
      >
        <button
          class="makeStyles-root-26 makeStyles-convertButton-10"
          data-testid="convertButton"
          disabled=""
          type="button"
        >
          Convert
        </button>
      </div>
    </div>
  </div>
</div>
`;
